### PR TITLE
SUSI bot - add support for advanced responses

### PR DESCRIPTION
**Pull request by hkedia321**
Improve susi bot by:
- adding support for table, anchor, answer type of responses
- restructuring actions processing logic by adding a `for loop`.

### CODE REVIEW

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 1]
```diff
@@ -41,9 +41,53 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                    # using markdown make a table
+                    for heading in columnsData:
+                        answer+="| "+heading+" "
+                    answer+="\n"
+                    for heading in columns:
+                        answer+="|"+"------"+" "
+                    answer+="\n"
+                    for item in (tableData):
+                        if(item[columns[0]]):
+                            column1=""
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: ideally, >3 indentations - and I usually call it - `something smells not clean`. Probably break it down at some point - return early, invert if, etc ?

Comment (hkedia321) [reply zulip_bots/zulip_bots/bots/susi/susi.py]: @tonythomas01 refactored the code to remove the if statements. It it okay now?

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 2]
```diff
@@ -41,9 +41,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+            answer = ""
+            for action in actions:
+                if action['type'] == 'answer':
+                    answer += action['expression'] + '\n'
+                elif action['type'] == 'anchor':
+                    text = action['text']
+                    link = action['link']
+                    answer += '[' + text + '](' + link+')' + '\n'
+                elif action['type'] == 'table':
+                    tableData = data['answers'][0]['data']
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: `table_data`

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 3]
```diff
@@ -41,9 +41,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                    msg = ""
+                    columns = list(action['columns'].keys())
+                    columnsData = list(action['columns'].values())
+                    # using markdown make a table
+                    for heading in columnsData:
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]:
```
for heading in list(action['columns'].values()):
...

```

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 4]
```diff
@@ -41,9 +41,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                        answer += "| " + heading + " "
+                    answer += "\n"
+                    for heading in columns:
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]:
```
for heading in list(action['columns'].keys())
...
```

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 5]
```diff
@@ -41,9 +41,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                        answer += "|" + "------" + " "
+                    answer += "\n"
+                    for item in (tableData):
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: unwanted paranthesis around `tableData` ?

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 6]
```diff
@@ -41,9 +41,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                        if (item[columns[0]]):
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: unwanted parenthesis

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 7]
```diff
@@ -41,9 +41,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                            column1 = item[columns[0]] if (type(item[columns[0]]) is str) else item[columns[0]][0]
+                            column2 = item[columns[1]] if (type(item[columns[1]]) is str) else item[columns[1]][0]
+                            column3 = item[columns[2]] if (type(item[columns[2]]) is str) else item[columns[2]][0]
+                            msg = "| " + column1+" | " + column2 + " | " + column3 + "| "
+                            answer = answer + msg + "\n"
+                    answer += "\n"
+        except Exception as e:
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: too broad exception here. What exceptions do you think would arise ?

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 8]
```diff
@@ -41,9 +41,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                    answer += "\n"
+                    for heading in columns:
+                        answer += "|" + "------" + " "
+                    answer += "\n"
+                    for item in (tableData):
+                        if (item[columns[0]]):
+                            column1 = item[columns[0]] if (type(item[columns[0]]) is str) else item[columns[0]][0]
+                            column2 = item[columns[1]] if (type(item[columns[1]]) is str) else item[columns[1]][0]
+                            column3 = item[columns[2]] if (type(item[columns[2]]) is str) else item[columns[2]][0]
+                            msg = "| " + column1+" | " + column2 + " | " + column3 + "| "
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: spacing around +

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 9]
```diff
@@ -41,9 +41,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                    # using markdown make a table
+                    for heading in columnsData:
+                        answer += "| " + heading + " "
+                    answer += "\n"
+                    for heading in columns:
+                        answer += "|" + "------" + " "
+                    answer += "\n"
+                    for item in (tableData):
+                        if (item[columns[0]]):
+                            column1 = item[columns[0]] if (type(item[columns[0]]) is str) else item[columns[0]][0]
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: hmm. the variable names seems to not make a lot of sense. Its minor, but it would be great to know what it is rather than a,b,c, a1, b1, c1 etc

File: zulip_bots/zulip_bots/bots/susi/test_susi.py [hunk 10]
```diff
@@ -39,7 +39,7 @@ def test_help(self) -> None:
 
     def test_issue(self) -> None:
         request = 'hi'
-        bot_response = 'Hello!'
+        bot_response = 'Hello!\n'
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/test_susi.py]: will this cover all these changes ?

Comment (hkedia321) [reply zulip_bots/zulip_bots/bots/susi/test_susi.py]: @tonythomas01 yes, I think this will be sufficient. The bot can reply with different answers as the answers come from the SUSI server. So this test case will just check if the bot APIs are working properly.

Comment (tonythomas01) [reply zulip_bots/zulip_bots/bots/susi/test_susi.py]: perfect.

File: zulip_bots/zulip_bots/bots/susi/susi.py (hunk 7 above)
Comment (hkedia321) [reply zulip_bots/zulip_bots/bots/susi/susi.py]: @tonythomas01 there can be many exceptions. Sometimes, the susi server might not reply with the expected object. This may result in exceptions as we are processing assuming a particular response object.

**Comment by hkedia321** [comment]
@tonythomas01 I made the suggested changes. Can you please review again.

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 11]
```diff
@@ -41,9 +41,38 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+                        if item[columns[0]]:
+                            column_first = item[columns[0]] if (type(item[columns[0]]) is str) else item[columns[0]][0]
+                            column_second = item[columns[1]] if (type(item[columns[1]]) is str) else item[columns[1]][0]
+                            column_third = item[columns[2]] if (type(item[columns[2]]) is str) else item[columns[2]][0]
+                            msg = "| " + column_first + " | " + column_second + " | " + column_third + "| "
+                            answer = answer + msg + "\n"
+                    answer += "\n"
+        except Exception as e:
             answer = "I don't understand. Can you rephrase?"
+            print(e)
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: is this print() intentional ?

Comment (hkedia321) [reply zulip_bots/zulip_bots/bots/susi/susi.py]: Yes, it was for debugging. Removed it now.

**Comment by hkedia321** [comment]
@timabbott @tonythomas01 Can you please review. Let me know if any changes are needed.

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 12]
```diff
@@ -41,8 +41,36 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
             return
         reply = requests.get("https://api.susi.ai/susi/chat.json", params=dict(q=msg))
         try:
-            answer = reply.json()['answers'][0]['actions'][0]['expression']
-        except Exception:
+            data = reply.json()
+            actions = data['answers'][0]['actions']
+            answer = ""
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: sorry for not finding this before, but I see you are using `"" ` now for strings. In other places, I see you arhe using `''` see line 39. This is inconsistent, and I would just put `''`, which is more clean.

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 13]
```diff
@@ -41,8 +41,36 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+            for action in actions:
+                if action['type'] == 'answer':
+                    answer += action['expression'] + '\n'
+                elif action['type'] == 'anchor':
+                    text = action['text']
+                    link = action['link']
+                    answer += '[' + text + '](' + link+')' + '\n'
+                elif action['type'] == 'table':
+                    table_data = data['answers'][0]['data']
+                    msg = ""
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: same here.

**Comment by hkedia321** [comment]
@tonythomas01 Changed all `""` to `''` :+1:

File: zulip_bots/zulip_bots/bots/susi/susi.py [hunk 14]
```diff
@@ -39,11 +39,39 @@ def handle_message(self, message: Dict[str, str], bot_handler: Any) -> None:
...
+            data = reply.json()
+            actions = data['answers'][0]['actions']
+            answer = ''
+            for action in actions:
+                if action['type'] == 'answer':
+                    answer += action['expression'] + '\n'
+                elif action['type'] == 'anchor':
+                    text = action['text']
+                    link = action['link']
+                    answer += '[' + text + '](' + link+')' + '\n'
```
Comment (tonythomas01) [inline zulip_bots/zulip_bots/bots/susi/susi.py]: missing space before binary operator here.

Comment (tonythomas01) [reply zulip_bots/zulip_bots/bots/susi/susi.py]: minor, but I think this should be fixed eventually as well.

Comment (hkedia321) [reply zulip_bots/zulip_bots/bots/susi/susi.py]: @tonythomas01 fixed. thanks for pointing out. :)

**Comment by showell** [comment]
This LGTM, but since this function does quite a bit of stuff, I think we want to add some test coverage before merging.  @hkedia321 Do you have time to add tests or should we have somebody else take over the PR?

Another thing to consider here is adding some helpers so that we're not doing so much string concatenation to build up the markdown.  I'm not concerned about the performance; it's just that the code might be nicer to read with things like `markdown_link(text=..., link=...)` and `markdown_table_header(...)`.

**Comment by hkedia321** [comment]
Sure, I will add test coverage and helper functions. @showell
